### DISCUSSION TITLE
Conditional activation depending on 3rd-party plugin parameter

### DISCUSSION THREAD

**Discussion by tanders** [helgoboss/helgobox#371] [Q&A]
Hi,

I would like to start with some praise of what ReaLearn offers in German: Tolle Arbeit! Die gebotene Flexibilität ist wirklich sehr schön. Glückwunsch! 

I already have a standard DAW control surface for handling standard workflow. I would like to use ReaLearn for more tricky issues. Specifically, I would like to use it for controlling parameters of plugins that happen to change their set of parameters on the fly, as is the case with the plugin family by Melda Productions.

These plugins feature (i) parameters definable by users (called multi parameters, basically the same as what is called macro controllers elsewhere, but rather flexible) and (ii) some notion of quasi meta presets called devices (a set of multi parameters + GUI). Devices and thus the set of available (multi) parameters for a given plugin can change on the fly.

Now, what I would like to do with ReaLearn is specifying mappings that depend on the currently chosen device of a plugin. I already figured out how to change the selected device by controlling the respective `Device` target parameter with some increment/decrement buttons. I also figured out how to (de)activate mapping groups by mapping the same increment/decrement buttons to one of ReaLearn's own parameters and then using that with the `When bank selected` activation condition. So, this setup basically does what I want, but it works only for a short time after setting it up and then soon breaks. 

One of the problems causing it to break is that the two mappings of the increment/decrement buttons (one for the Melda `Device` target parameter and the other for ReaLearn's own parameters) run soon out of sync. One reason for that might be that if the target track is set to `<Selected>`, this setting is seemingly only respected by one of the two mappings. The mapping to ReaLearn's own parameters seems to ignore that setting and be always active. 

Another problem is that users can define new devices with these flexible plugins (at least for some of them), and that can scramble quasi the index (bank number) of devices. Each device has a name, and this name happens to be also shown as part of the parameter value by ReaLearn under the target mapping, but this is obviously not used currently for the conditional activation.

So, I have multiple questions (or possibly feature requests) here.

1) Is it perhaps possible to define a conditional activation that depends directly on the value of some parameter of some third-party plugin? (to avoid that some internal parameter runs out of sync with an external parameter.)

2) _[Low priority]_ Is it perhaps possible to define a conditional activation that is not just a numeric condition, but depends on some associated information of this plugin value (I actually have no idea how this metadata is handled by plugins, but I can see that ReaLearn already shows that information).

3) In case neither request 1 nor 2 are possible at the moment, is there perhaps another workaround to avoid that multiple mappings of the same sources to different targets ran out of sync? I guess, as a last resort I can always simply define separate ReaLearn mapping presets for separate devices of a single plugin and then just load whatever I need. That is still a customisation I cannot do with my general DAW controller, because there I cannot define and switch between multiple mappings for a single plugin. However, I might end up with very many presets in total then, so some way to select them that goes beyond a menu (e.g., a standard file selection dialog) would be helpful at some later stage. :) 

Apologies for this rather long message. Thanks a lot again for ReaLearn! 

PS: While Melda Production also offers a bundle of free plugins, unfortunately the notion of devices is only supported by paid plugins.

**Comment by helgoboss** [comment]
Let me quickly answer question 1 for now: It's already possible given you have the third-party plug-in in the same FX chain:

1. Display ReaLearn's parameter list by pressing "UI" button.
2. Move a yet non-used main compartment parameter, e.g. "Main p1".
3. Press Param > Parameter modulation/MIDI link
4. Enable "Link from MIDI or FX parameter"
5. Choose the parameter of the other plug-in.

This makes sure the third-party plug-in parameter and conditional activation stay in sync. It's a one-way sync only ... uni-directional (but you can set up ReaLearn to control the third-party plug-in parameter so I don't think this is a showstopper). You could also choose the other direction.

Another way would be to either use 2 mappings with the same source that control both ReaLearn's parameter and the third-party plug-in parameter - and make sure that you only use ReaLearn to modify them. A variation of this is to use the new "Group interaction" feature (ReaLearn v2.10.0-pre.2) with the "Same target value" option. But that's not much different, you just don't have to define the source twice.

**Reply by tanders** [reply]
Thanks for coming back so quickly.

> question 1: It's already possible

This already seems to work in principle, great! 

The problem I still have with this though is that it is a bit difficult to work out which `Main p1` value now corresponds to which Melda plugin `device` index. Say, I have a Melda plugin that happens to provides `device` indices (integers) in the interval `[0, 21]`. With the `Link from MIDI or FX parameter` applied, `p1` now moves in the interval `[0, 0.625]`. Of course I can now do the math to figure out which `p1` value corresponds to which `device` index to set the correct value for the conditional activation, but it is a but cumbersome. ;)  I tried using the `offset` and `scale` (in the Parameter modulation/MIDI link dialog) for compensation, but seemingly I don't quite understand them.

>  given you have the third-party plug-in in the same FX chain

So, in case the same plugin is used multiple times, then a separate ReaLearn instance with the respective preset needs to be preset in the same FX chain? It will not cause problems if they use the same source (controller) of the mappings? The workflow using the monitoring FX chain for automatically loaded mappings is not yet supported?  NOTE: This is all fine and ReaLearn is great, I just want to make sure I understand correctly. :) 

Thanks again!

**Reply by tanders** [reply]
EDIT: To clarify, my question here is how to find out the relevant program bank numbers that I need to use in the activation conditions if the relevant parameter(`Main p1) is linked to a 3rd-party plugin parameter and this results in a parameter value range as described above. Perhaps I am simply misunderstanding or missing something.

**Reply by helgoboss** [reply]
> The problem I still have with this though is that it is a bit difficult to work out which `Main p1` value now corresponds to which Melda plugin `device` index. Say, I have a Melda plugin that happens to provides `device` indices (integers) in the interval `[0, 21]`. With the `Link from MIDI or FX parameter` applied, `p1` now moves in the interval `[0, 0.625]`. Of course I can now do the math to figure out which `p1` value corresponds to which `device` index to set the correct value for the conditional activation, but it is a but cumbersome. ;) I tried using the `offset` and `scale` (in the Parameter modulation/MIDI link dialog) for compensation, but seemingly I don't quite understand them.
> EDIT: To clarify, my question here is how to find out the relevant program bank numbers that I need to use in the activation conditions if the relevant parameter(`Main p1) is linked to a 3rd-party plugin parameter and this results in a parameter value range as described above. Perhaps I am simply misunderstanding or missing something.

I get it. This could be improved with #209. When implemented, this lets you enter a discrete number of values that for each internal ReaLearn parameter (in your case 22). I could use this information to fill the list of banks accordingly. Vote for this issue if this would help you.

> So, in case the same plugin is used multiple times, then a separate ReaLearn instance with the respective preset needs to be preset in the same FX chain? It will not cause problems if they use the same source (controller) of the mappings? The workflow using the monitoring FX chain for automatically loaded mappings is not yet supported? NOTE: This is all fine and ReaLearn is great, I just want to make sure I understand correctly. :)

Sorry, I think I don't understand any of these questions. Can you rephrase?
